src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs

```diff
@@ -130,17 +130,33 @@ private void MarkLocalsUnassigned()
             for (int i = 0; i < nextVariableSlot; i++)
             {
                 var symbol = variableBySlot[i].Symbol;
-                var local = symbol as LocalSymbol;
-                if ((object)local != null && !local.IsConst)
-                {
-                    SetSlotState(i, false);
-                    continue;
-                }
 
-                var parameter = symbol as ParameterSymbol;
-                if ((object)parameter != null)
+                if ((object)symbol != null)
                 {
-                    SetSlotState(i, false);
+                    switch (symbol.Kind)
+                    {
+                        case SymbolKind.Local:
+                            if (!((LocalSymbol)symbol).IsConst)
+                            {
+                                SetSlotState(i, false);
+                            }
+                            break;
+
+                        case SymbolKind.Parameter:
+                            SetSlotState(i, false);
+                            break;
+
+                        case SymbolKind.Field:
+                            if (!((FieldSymbol)symbol).IsConst)
+                            {
+                                SetSlotState(i, false);
+                            }
```

Comment:
Do we care whether it is a static/instance field or whether container is a class/struct?